### DISCUSSION TITLE
How do I re-authenticate via API using useRuntimeConfig variables?

### DISCUSSION THREAD

**Discussion by danielvlla** [nuxt/nuxt#21695] [Questions] [answered]
I have a Vue component which is used to obtain data from my API. I use Pinia to save the JWT token, however, in some instances, the API needs to reauthenticate and on the client side, the Authentication failed error is thrown, and the `config.apiUsername` and `config.apiPassword` are undefined (hence why the error is thrown). These items are defined in the `nuxt.config.ts` file under `runTimeConfig` as follows:

```
  runtimeConfig: {
    apiUsername: 'abc',
    apiPassword: 'password123',
    public: {
      apiURL: "XYZX",
      siteUrl: "XYZ
    },
  },

```

I am using the below in a <script setup> tag

```
    const authenticate = async () => {
        const config = useRuntimeConfig()

        const formData = new FormData()
        formData.append('username', config.apiUsername)
        formData.append('password', config.apiPassword)

        const { data: token } = await useFetch(
            config.public.apiURL + 'api/token',
            {
                method: 'POST',
                body: formData,
            }
        )

        if (token && token.value && token.value.access_token) {
            $store.set(token.value.access_token)
        } else {
            throw new Error('Authentication failed')
        }
    }

    const fetchData = async () => {     
        const config = useRuntimeConfig()

        const { error, data } = await useFetch(
            config.public.apiURL + 'api/stations',
            {
                headers: {
                    Authorization: `Bearer ${$store.jwt}`,
                },
            }
        )

        if (error && error.value && error.value.statusCode === 401) {
            await authenticate()
            await fetchData()
        }

        error.value = error.value
        records.value = data.value.records
        last_update.value = data.value.last_update
    }
    
    if ($store.jwt) {
        await fetchData()
    } else {
        await authenticate()
        await fetchData()
    }
```

Is the above good data flow in Nuxt or am I missing something? Also if I move `await fetchData()` outside of the if-else block, I get the following error:

_A composable that requires access to the Nuxt instance was called outside of a plugin, Nuxt hook, Nuxt middleware, or Vue setup function. This is probably not a Nuxt bug. Find out more at `https://nuxt.com/docs/guide/concepts/auto-imports#using-vue-and-nuxt-compos_

**Comment by danielvlla** [comment]
PS - When this happens, a refresh of the page fixes the problem, however, I do not understand why a refresh is needed for the problem to be fixed.

**Comment by TheAlexLichter** [comment, marked as answer]
Because when you refresh, the `runtimeConfig` values that are not `public` are available during SSR (on the server-side).
Because they are not public, you can't re-authenticate via client.

I think a good option would be a reauthenticate endpoint via Nitro that you could call. That endpoint can use the apiUsername and password without leaking it to the client side then ☺️

**Reply by danielvlla** [reply]
Thank you for sharing, I will try it out 🙏🏼
